File: src/main/scala/com/example/igor/amazingview/AmazingViewHolder.scala
package com.example.igor.amazingview

import android.view.View
import rx.functions.Action1
import android.view.View.OnClickListener

abstract class AmazingViewHolder(itemView: View) {

  def getItemView : View = itemView

  protected def onClickCallback: String

  def setOnClickAction(action: Action1[String]): AmazingViewHolder = {
    val listener = new OnClickListener {
      override def onClick(v: View): Unit = action.call(onClickCallback)
    }
    itemView.setOnClickListener(listener)
    this
  }
}

File: src/main/java/com/example/igor/amazingview/AmazingViewHolder.java
package com.example.igor.amazingview;

import android.view.View;
import rx.functions.Action1;

public abstract class AmazingViewHolder {
    public final View itemView;

    public AmazingViewHolder(View itemView) {
        this.itemView = itemView;
    }

    protected abstract String onClickCallback();

    public AmazingViewHolder setOnClickAction(final Action1<String> action) {
        itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                action.call(onClickCallback());
            }
        });
        return this;
    }
}
